python/adjoint/objective.py

```diff
@@ -21,152 +41,231 @@ def register_monitors(self, frequencies):
+          `forward` and should have a value of either 0 or 1.
+    """
+    def __init__(
+        self,
+        sim,
+        volume,
+        mode,
+        forward=True,
+        kpoint_func=None,
+        kpoint_func_overlap_idx=0,
+        **kwargs,
+    ):
+        super().__init__(sim)
+        if kpoint_func_overlap_idx not in [0, 1]:
+            raise ValueError(
+                f'`kpoint_func_overlap_idx` should be either 0 or 1, but got {kpoint_func_overlap_idx}.'
+            )
         self.volume = volume
         self.mode = mode
         self.forward = forward
-        self.normal_direction = None
         self.kpoint_func = kpoint_func
-        self.eval = None
-        self.EigenMode_kwargs = kwargs
+        self.kpoint_func_overlap_idx = kpoint_func_overlap_idx
+        self.eigenmode_kwargs = kwargs
+        self._monitor = None
+        self._direction_vector = None
+        self._cscale = None
 
     def register_monitors(self, frequencies):
-        self.frequencies = np.asarray(frequencies)
-        self.monitor = self.sim.add_mode_monitor(
+        self._frequencies = np.asarray(frequencies)
+        self._monitor = self.sim.add_mode_monitor(
             frequencies,
             mp.ModeRegion(center=self.volume.center, size=self.volume.size),
             yee_grid=True,
         )
-        self.normal_direction = self.monitor.normal_direction
-        return self.monitor
+        if self._monitor.normal_direction == 0:
+            direction_vector = mp.Vector3(x=1)
+        elif self._monitor.normal_direction == 1:
+            direction_vector = mp.Vector3(y=1)
+        elif self._monitor.normal_direction == 2:
+            direction_vector = mp.Vector3(z=1)
+        self._direction_vector = direction_vector if self.forward else -1 * direction_vector
+        return self._monitor
 
     def place_adjoint_source(self, dJ):
         dJ = np.atleast_1d(dJ)
-        dt = self.sim.fields.dt
-        direction_scalar = -1 if self.forward else 1
-        if self.kpoint_func is None:
-            if self.normal_direction == 0:
-                k0 = direction_scalar * mp.Vector3(x=1)
-            elif self.normal_direction == 1:
-                k0 = direction_scalar * mp.Vector3(y=1)
-            elif self.normal_direction == 2:
-                k0 = direction_scalar * mp.Vector3(z=1)
-        else:
-            k0 = direction_scalar * self.kpoint_func(self.time_src.frequency,
-                                                     1)
         if dJ.ndim == 2:
             dJ = np.sum(dJ, axis=1)
-        da_dE = 0.5 * self.cscale  # scalar popping out of derivative
+        time_src = self._create_time_profile()
+        da_dE = 0.5 * self._cscale
+        scale = self._adj_src_scale()
 
-        scale = adj_src_scale(self, dt)
+        if self.kpoint_func:
+            eig_kpoint = self.kpoint_func(time_src.frequency, self.mode)
+        else:
+            eig_kpoint = self._direction_vector
```

Comment:
This isn't quite the right guess, since it has the wrong units.  It would be better to use mpb.cpp's guess for this, which is obtained by passing `(0.0,0.0,0.0)` for the k-point.  To tell it to flip the sign, we could pass `(-0.0,-0.0,-0.0)` and use the [sign of zero](https://people.freebsd.org/~das/kahan86branch.pdf):
```diff
diff --git a/src/mpb.cpp b/src/mpb.cpp
index c41e3c2a..5a68ca7b 100644
--- a/src/mpb.cpp
+++ b/src/mpb.cpp
@@ -484,7 +484,7 @@ void *fields::get_eigenmode(double frequency, direction d, const volume where, c
   // which we automatically pick if kmatch == 0.
   if (match_frequency && kmatch == 0) {
     vec cen = eig_vol.center();
-    kmatch = frequency * sqrt(real(get_eps(cen, frequency)) * real(get_mu(cen, frequency)));
+    kmatch = copysign(frequency * sqrt(real(get_eps(cen, frequency)) * real(get_mu(cen, frequency))), kmatch);
     if (d == NO_DIRECTION) {
       for (int i = 0; i < 3; ++i)
         k[i] = dot_product(R[i], kdir) * kmatch; // kdir*kmatch in reciprocal basis
```

Reply:
Looks like the sign is being lost in kmatch, so you'd have to look at `_kpoint.in_direction(d)` as the second argument of `copysign`